File: Lab06Cs/Lab06Monitoring/HumiditySensor.cs
using System;

namespace Lab06Monitoring
{
    /// <summary>
    /// Датчик влажности.
    ///
    /// ЛР-6:
    ///  * второй пример наследования;
    ///  * переопределяем виртуальные методы;
    ///  * в Calibrate() НЕ вызываем базовую реализацию.
    /// </summary>
    public class HumiditySensor : Sensor
    {
        public float MinThreshold { get; }
        public float MaxThreshold { get; }

        private static readonly Random _rnd = new();

        public HumiditySensor(string id, string name, float min, float max)
            : base(id, name, "%")
        {
            if (min >= max)
                throw new ArgumentException("min must be < max", nameof(min));

            MinThreshold = min;
            MaxThreshold = max;
        }

        public override float ReadValue()
        {
            LastValue = (float)(_rnd.NextDouble() * (MaxThreshold - MinThreshold + 40) + (MinThreshold - 20));
            Console.WriteLine($"HumiditySensor '{Name}' read: {LastValue:F2} {Unit}");
            return LastValue;
        }

        public override bool CheckStatus()
        {
            bool ok = LastValue >= MinThreshold && LastValue <= MaxThreshold;
            Console.WriteLine($"HumiditySensor '{Name}' status: {(ok ? "OK" : "OUT OF RANGE")}");
            return ok;
        }

        public override void Calibrate()
        {
            Console.WriteLine($"HumiditySensor '{Name}': fast calibration without base call");
        }

        public override string Describe()
        {
            return $"HumiditySensor[{Id}] '{Name}', range = {MinThreshold}..{MaxThreshold} {Unit}";
        }

        public override object Clone()
        {
            return new HumiditySensor(Id, Name, MinThreshold, MaxThreshold)
            {
                LastValue = this.LastValue,
                Status = this.Status
            };
        }
    }
}

File: Lab08Cpp/HumiditySensor.cpp
#include "HumiditySensor.hpp"
#include <stdexcept>
#include <iomanip>

HumiditySensor::HumiditySensor(const std::string& id,
                               const std::string& name,
                               float min,
                               float max)
    : Sensor(id, name, "%")
    , m_min(min)
    , m_max(max)
{
    if (min >= max)
    {
        throw std::invalid_argument("HumiditySensor: min threshold must be < max threshold");
    }
}

float HumiditySensor::ReadValue()
{
    static std::mt19937 gen{std::random_device{}()};
    std::uniform_real_distribution<float> dist(m_min - 20.0f, m_max + 20.0f);

    m_lastValue = dist(gen);
    std::cout << "HumiditySensor '" << m_name
              << "' read value: " << std::fixed << std::setprecision(2)
              << m_lastValue << " " << m_unit << std::endl;
    return m_lastValue;
}

bool HumiditySensor::CheckStatus() const
{
    bool ok = (m_lastValue >= m_min && m_lastValue <= m_max);
    std::cout << "HumiditySensor '" << m_name << "' status: "
              << (ok ? "OK" : "OUT OF RANGE") << std::endl;
    return ok;
}

void HumiditySensor::Calibrate()
{
    // ЛР-6: переопределение метода БЕЗ вызова базовой реализации.
    std::cout << "HumiditySensor '" << m_name << "': fast calibration without base call" << std::endl;
}

std::string HumiditySensor::Describe() const
{
    return "HumiditySensor[" + m_id + "] '" + m_name + "', range = " +
           std::to_string(m_min) + ".." + std::to_string(m_max) + " " + m_unit;
}

Sensor* HumiditySensor::Clone() const
{
    return new HumiditySensor(*this);
}
